### PR TITLE
DRS filesystem: Read support

### LINKED ISSUE #4292: Retrieve contents for DOS URIs
Today, there's no way to use certain language functions with DOS URI file types (such as read_lines(), read_string() etc). This makes it hard to use existing workflows with DOS inputs.

AC: Add support for being able to get file contents for DOS URIs as a part of the DOS filesystem provider in Cromwell. Ensure that this workflows for both workflow level functions and task level functions. For example:

```
workflow w {
File f = "dos://uuid"
String x = read_string(f)

call t { 
  input: 
     f = f
     x = x
}
}

task t {
  String x
  File f
  String y = read_string(f)
 
  command {
    if [[ ${x} != ${y} ]]; then
      exit 1
    fi
  }
  runtime {
    docker: "ubuntu"
  }
}
```

### PR DESCRIPTION
**Pull request by salonishah11**
This PR adds support for `read_lines` and `read_string` for DRS inputs.

Closes #4292

### CODE REVIEW

File: cloud-nio/cloud-nio-impl-drs/src/main/scala/cloud/nio/impl/drs/DrsCloudNioFileProvider.scala [hunk 1]
```diff
@@ -36,30 +44,85 @@ class DrsCloudNioFileProvider(config: Config, scheme: String, drsPathResolver: D
...
+
+    schemeUrlOption match {
+      case Some(schemeUrl) => schemeUrl.url
+      case None => throw UrlNotFoundException(drsPath, scheme)
+    }
+  }
+
+
+  private def gcsInputStream(bucket: String, fileToBeLocalized: String, serviceAccount: String): ReadChannel = {
+    val credentials = GoogleCredentials.fromStream(new ByteArrayInputStream(serviceAccount.getBytes()))
```
Comment (danbills) [inline cloud-nio/cloud-nio-impl-drs/src/main/scala/cloud/nio/impl/drs/DrsCloudNioFileProvider.scala]: let's put these 2 lines at the end of the world.  I'm submitting a PR with some suggestions

File: cloud-nio/cloud-nio-impl-drs/src/main/scala/cloud/nio/impl/drs/DrsCloudNioFileProvider.scala [hunk 2]
```diff
@@ -15,13 +19,17 @@ class DrsCloudNioFileProvider(config: Config, scheme: String, drsPathResolver: D
 
   private lazy val httpClientBuilder = HttpClientBuilder.create()
 
+  private val GcsScheme = "gs"
+
+
   private def getDrsPath(cloudHost: String, cloudPath: String): String = s"$scheme://$cloudHost/$cloudPath"
 
+
```
Comment (mcovarr) [inline cloud-nio/cloud-nio-impl-drs/src/main/scala/cloud/nio/impl/drs/DrsCloudNioFileProvider.scala]: TOL extra unnecessary vertical space

File: cloud-nio/cloud-nio-impl-drs/src/main/scala/cloud/nio/impl/drs/DrsCloudNioFileProvider.scala [hunk 3]
```diff
@@ -36,30 +44,85 @@ class DrsCloudNioFileProvider(config: Config, scheme: String, drsPathResolver: D
   }
 
 
+  private def extractUrlForScheme(drsPath: String, urlArray: Array[Url], scheme: String): String = {
+    val schemeUrlOption = urlArray.find(u => u.url.startsWith(scheme))
```
Comment (mcovarr) [inline cloud-nio/cloud-nio-impl-drs/src/main/scala/cloud/nio/impl/drs/DrsCloudNioFileProvider.scala]: TOL 

```scala
find(_.url.startsWith(scheme))
```

File: cloud-nio/cloud-nio-impl-drs/src/main/scala/cloud/nio/impl/drs/DrsCloudNioFileProvider.scala [hunk 4]
```diff
@@ -36,30 +44,85 @@ class DrsCloudNioFileProvider(config: Config, scheme: String, drsPathResolver: D
...
+    blob.reader()
+  }
+
+
+  private def inputStream(url: String, serviceAccount: String, scheme: String): ReadChannel = {
+    scheme match {
+      case GcsScheme => {
+        val urlArray = url.replace(s"$scheme://", "").split("/", 2)
+        val fileToBeLocalized = urlArray(1)
+        val bucket = urlArray(0)
```
Comment (mcovarr) [inline cloud-nio/cloud-nio-impl-drs/src/main/scala/cloud/nio/impl/drs/DrsCloudNioFileProvider.scala]: TOL

```
val Array(fileToBeLocalized, bucket) = url.replace(s"$scheme://", "").split("/", 2)
```

File: cloud-nio/cloud-nio-impl-drs/src/main/scala/cloud/nio/impl/drs/DrsCloudNioFileProvider.scala (hunk 3 above)
Comment (salonishah11) [reply cloud-nio/cloud-nio-impl-drs/src/main/scala/cloud/nio/impl/drs/DrsCloudNioFileProvider.scala]: 👍

File: cloud-nio/cloud-nio-impl-drs/src/main/scala/cloud/nio/impl/drs/DrsCloudNioFileProvider.scala (hunk 4 above)
Comment (salonishah11) [reply cloud-nio/cloud-nio-impl-drs/src/main/scala/cloud/nio/impl/drs/DrsCloudNioFileProvider.scala]: 👍

**Comment by danbills** [comment]
I think we should close this as it is under heavy construction, and re-reviewed once the dust has cleared

**Pull request closed without merging**